designs/node-repair.md

```diff
@@ -0,0 +1,93 @@
+#  Karpenter Node Auto Repair 
+
+## Problem Statement
+
+Karpenter users are observing that occasionally nodes can face failure mode that prevent them to be healthy in a Kubernetes clusters. These nodes stay in unhealthy state for a period of time without any actions, and in some cases forever with manual intervention. Karpenter continues to simulate pods to these nodes, but the Kubernetes scheduler won't schedule to them. This will leave pods unscheduled and cause downtime for applications. Customers have the need for karpenter to repair unhealthy nodes. The karpenter team has seen several failure mode such as disk pressure, memory pressure, networking outage etc. that can cause nodes to get into an unhealthy state Here are issues related to the topic:
+
+* Mega Issue: https://github.com/kubernetes-sigs/karpenter/issues/750
+    * Related (Unreachable): https://github.com/aws/karpenter-provider-aws/issues/2570
+    * Related (Remove by taints): https://github.com/aws/karpenter-provider-aws/issues/2544
+    * Related (Known resource are not registered) Fixed by v0.28.0: https://github.com/aws/karpenter-provider-aws/issues/3794
+    * Related (Stuck on NotReady): https://github.com/aws/karpenter-provider-aws/issues/2439
+
+#### Out of scope
+
+The scope of the solution suggested below will be an opinionated method that will not give customer configurability. The team does not have enough data to determine the right level of configuration surface that users will utilize. The implementation will not consider budgets, API surface for repairing unhealthy nodes, or option for nodes to be rebooted instead of replaced. The solution will also not give any Karpenter based approach to enable AZ resiliency. **The feature will be gated under an alpha feature flag to allow for additional feedback from customers, and make subsequent changes in the future.**
+
+### Option  1 (recommended) : Unhealthy condition set cloud provider interface  
+
+```
+type RepairPolicy struct {
+    // Type of unhealthy state that is found on the node
+    Type metav1.ConditionType 
+    // Status condition of when a node is unhealthy
+    Status metav1.ConditionStatus
+    // RepairAfter is the duration the controller will wait
+    // before attempting to terminate nodes that are marked for repair.
+    TolerationDuration time.Duration
+}
+
+type CloudProvider interface {
+  ...
+    // RepairPolicy is for CloudProviders to define a set Unhealthy condition for Karpenter 
+    // to monitor on the node. Customer will need 
+    RepairPolicy() []v1.RepairPolicy
```

Comment:
Is there any way for a user to block repair from happening after enabling this feature? I'm assuming that `karpenter.sh/do-not-disrupt` won't work here but what about if I'm a user and I want to debug?

Reply:
We could add it a block just for that annotation, but it might be best for users to disable the feature to investigate. They would also have metrics and alarming to be able to give a better insight

Reply:
I think we have a similar question with expiration -- should we allow users to add `karpenter.sh/do-not-disrupt` to block the deletion of a node when we don't support any budget for the disruption type -- maybe more generally -- should we support budgets for "forceful" disruption mechanism where customers don't want application devs to get in the way but they are fine getting in their own way

Reply:
I think for the time being we should not added it, and see if that is something that customers need